test/selectors_test.rb

```diff
@@ -0,0 +1,154 @@
+                      value: { value: "first-line" }
+                    ]
+                  ],
+                  []
+                ]
+              ]
+            ]
+          ]
+        end
+      end
+
+      it "parses a complex selector" do
+        actual = parse_selectors("section>table")
+
+        assert_pattern do
+          actual => [
+            Selectors::ComplexSelector[
+              left: Selectors::TypeSelector[value: { name: { value: "section" } }],
+              combinator: { value: { value: ">" } },
+              right: Selectors::TypeSelector[value: { name: { value: "table" } }]
+            ]
+          ]
+        end
+      end
+
+      it "parses a complex selector with many selectors" do
+        actual = parse_selectors("section>table>tr")
+
+        assert_pattern do
+          actual => [
+            Selectors::ComplexSelector[
+              left: Selectors::TypeSelector[value: { name: { value: "section" } }],
+              combinator: { value: { value: ">" } },
+              right: Selectors::ComplexSelector[
+                left: Selectors::TypeSelector[value: { name: { value: "table" } }],
+                combinator: { value: { value: ">" } },
+                right: Selectors::TypeSelector[value: { name: { value: "tr" } }]
+              ]
+            ]
+          ]
+        end
+      end
+
+      it "parses a complex selector with whitespace" do
+        actual = parse_selectors("section > table")
+
+        assert_pattern do
+          actual => [
+            Selectors::ComplexSelector[
+              left: Selectors::TypeSelector[value: { name: { value: "section" } }],
+              combinator: { value: { value: ">" } },
+              right: Selectors::TypeSelector[value: { name: { value: "table" } }],
+            ]
+          ]
+        end
+      end
+
+      it "parses a complex selector with implicit descendant combinator" do
+        actual = parse_selectors("section table")
+
+        assert_pattern do
+          actual => [
+            Selectors::ComplexSelector[
+              left: Selectors::TypeSelector[value: { name: { value: "section" } }],
+              combinator: nil,
+              right: Selectors::TypeSelector[value: { name: { value: "table" } }],
+            ]
+          ]
+        end
+      end
+
+      it "parses a complex complex selector" do
+        actual = parse_selectors("section > table tr")
+
+        assert_pattern do
+          actual => [
+            Selectors::ComplexSelector[
+              left: Selectors::TypeSelector[value: { name: { value: "section" } }],
+              combinator: { value: { value: ">" } },
+              right: Selectors::ComplexSelector[
+                left: Selectors::TypeSelector[value: { name: { value: "table" } }],
+                combinator: nil,
+                right: Selectors::TypeSelector[value: { name: { value: "tr" } }]
+              ]
+            ]
+          ]
+        end
```

Comment:
After sleeping on it, I don't like this recursive structure. The [spec defines a `complex-selector`](https://www.w3.org/TR/selectors-4/#typedef-complex-selector) as:

```
<complex-selector> = <compound-selector> [ <combinator>? <compound-selector> ]*
```

which would imply a preferred AST like:

```ruby
          assert_pattern do
            actual => [
              Selectors::ComplexSelector[
                [
                  Selectors::TypeSelector[value: { name: { value: "section" } }],
                  Selectors::Combinator[value: { value: ">" }],
                  Selectors::TypeSelector[value: { name: { value: "table" } }],
                  Selectors::TypeSelector[value: { name: { value: "tr" } }]
                ]
              ]
            ]
          end
```

I'd like to change this, either in this PR or in the next one. @kddnewton thoughts?

Reply:
I've got a commit the converts to this flattened AST ... let me know what you want to do!

Reply:
Flatter is better for sure

Reply:
OK, pushed a commit that flattens this AST, LMK what you think.